packages/react-async/src/component.tsx

```diff
@@ -1,135 +1,330 @@
-        defer={defer}
-        renderLoading={renderLoading}
-        render={Component =>
-          Component ? <Component {...componentProps} /> : null
+      <>
+        {contentMarkup}
+        {loadingMarkup}
+      </>
+    );
+  }
+
+  Async.displayName = `Async(${componentName})`;
+
+  function usePreload(props: PreloadOptions) {
+    const {load} = useAsync(resolver, {
+      assets: AssetTiming.NextPage,
+    });
+
+    const customPreload = useCustomPreload(props);
+
+    return useCallback(
+      () => {
+        load();
+
+        if (customPreload) {
+          customPreload();
         }
-        {...asyncProps}
-      />
+      },
+      [load, customPreload],
     );
   }
 
-  function Preload(props: PreloadProps & ConstantProps) {
-    const [componentProps, asyncProps] = splitProps(props);
+  function usePrefetch(props: PrefetchOptions) {
+    const {load} = useAsync(resolver, {
+      assets: AssetTiming.NextPage,
+    });
 
-    return (
-      <>
-        {renderPreload(componentProps)}
-        <Async defer={DeferTiming.Idle} load={load} {...asyncProps} />
-      </>
+    const customPrefetch = useCustomPrefetch(props);
+
+    return useCallback(
+      () => {
+        load();
+
+        if (customPrefetch) {
+          customPrefetch();
+        }
+      },
+      [load, customPrefetch],
     );
   }
 
-  function Prefetch(props: PrefetchProps & ConstantProps) {
-    const [componentProps, asyncProps] = splitProps(props);
+  function useKeepFresh(props: KeepFreshOptions) {
+    const {load} = useAsync(resolver, {
+      assets: AssetTiming.NextPage,
+    });
 
-    return (
-      <>
-        {renderPrefetch(componentProps)}
-        <Async defer={DeferTiming.Mount} load={load} {...asyncProps} />
-      </>
+    const customKeepFresh = useCustomKeepFresh(props);
+
+    return useCallback(
+      () => {
+        load();
+
+        if (customKeepFresh) {
+          customKeepFresh();
+        }
+      },
+      [load, customKeepFresh],
     );
   }
 
-  function KeepFresh(props: KeepFreshProps & ConstantProps) {
-    const [componentProps, asyncProps] = splitProps(props);
+  function Preload(options: PreloadOptions) {
+    useIdleCallback(usePreload(options));
+    return null;
+  }
 
-    return (
-      <>
-        {renderKeepFresh(componentProps)}
-        <Async defer={DeferTiming.Idle} load={load} {...asyncProps} />
-      </>
+  Preload.displayName = `Async.Preload(${displayName})`;
+
+  function Prefetch(options: PrefetchOptions) {
+    const prefetch = usePrefetch(options);
+
+    useEffect(
+      () => {
+        prefetch();
+      },
+      [prefetch],
     );
+
+    return null;
+  }
+
+  Prefetch.displayName = `Async.Prefetch(${displayName})`;
+
+  function KeepFresh(options: KeepFreshOptions) {
+    useIdleCallback(useKeepFresh(options));
+    return null;
   }
 
+  KeepFresh.displayName = `Async.KeepFresh(${displayName})`;
+
   // Once we upgrade past TS 3.1, this will no longer be necessary,
```

Comment:
I know this isn't added in this PR, but we are past 3.1 in web, is this still relevant here?